pkg/codegen/importer/hcl2.go

```diff
@@ -365,11 +366,222 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
+}
+
+// reduceUnionType reduces the given union type to a simpler type that potentially matches the value.
+// When the value type is primitive, choose the first element type of the union elements that is of the same type.
+// When the value is an object, use the discriminator to choose the element type.
+func reduceUnionType(schemaUnion *schema.UnionType, value resource.PropertyValue) schema.Type {
+	switch {
+	case value.IsObject():
+		// return the first element type that matches structurally fits the value
+		findBestFitType := func() schema.Type {
+			for _, t := range schemaUnion.ElementTypes {
+				if valueStructurallyTypedAs(value, t) {
+					return t
+				}
+			}
+
+			// if we still couldn't find a type that fits the value
+			return nil
+		}
+
+		// If the value is an object, use the discriminator to choose the element type.
+		if schemaUnion.Discriminator == "" {
+			return findBestFitType()
+		}
+
+		obj := value.ObjectValue()
+		discriminatorValue, ok := obj[resource.PropertyKey(schemaUnion.Discriminator)]
+		if !ok {
+			// discriminator property is not present
+			// return the first type that fits the value
+			return findBestFitType()
+		}
+
+		if !discriminatorValue.IsString() {
+			// discriminator property value is not a string,
+			// so we can't select a type from the union mapping
+			return findBestFitType()
+		}
+
+		correspondingTypeToken, ok := schemaUnion.Mapping[discriminatorValue.StringValue()]
+		if !ok {
+			// discriminator property value is not a key in the union mapping,
+			return findBestFitType()
+		}
+
+		for _, elementType := range schemaUnion.ElementTypes {
+			// found the type token
+			// match it against the element type which should be an object
+			elementTypeObject, ok := codegen.UnwrapType(elementType).(*schema.ObjectType)
+			if ok {
+				elementTypeToken, parseError := tokens.ParseTypeToken(elementTypeObject.Token)
+				if parseError != nil {
+					continue
+				}
+
+				foundTypeToken, parseError := tokens.ParseTypeToken(correspondingTypeToken)
+
+				if parseError != nil {
+					continue
+				}
+
+				typeName := string(elementTypeToken.Name())
+				foundTypeName := string(foundTypeToken.Name())
+				if typeName == foundTypeName {
+					return elementTypeObject
+				}
+			}
+		}
+
+	case value.IsString():
+		// check to see if one of the element types in the union
+		// actually allows a string type and return it
+		for _, t := range schemaUnion.ElementTypes {
+			if t == schema.StringType {
+				return t
+			}
+		}
+
+	case value.IsBool():
+		// check to see if one of the element types in the union
+		// actually allows a bool type and return it
+		for _, t := range schemaUnion.ElementTypes {
+			if t == schema.BoolType {
+				return t
+			}
+		}
+
+	case value.IsNumber():
+		// check to see if one of the element types in the union
+		// actually allows a number type and return it
+		for _, t := range schemaUnion.ElementTypes {
+			if t == schema.NumberType {
+				return t
+			}
+		}
+	}
```

Comment:
Unions come in two types: discriminated objects and structural. We can simplify this (and handle the missing `case value.IsArray()`) by invoking `valueStructurallyTypedAs`:
```suggestion
	default:
		for _, t := range schemaUnion.ElementTypes {
			if valueStructurallyTypedAs(value, t) {
				return t
			}
		}
	}
```